File: samples/msgext-northwind-inventory-ts/src/messageExtensions/revenueSearchCommand.ts
import {
    CardFactory,
    TurnContext,
    MessagingExtensionQuery,
    MessagingExtensionResponse,
} from "botbuilder";
import { getProductsByRevenueRange } from "../northwindDB/products";
import cardHandler from "../adaptiveCards/cardHandler";

const COMMAND_ID = "revenueSearch";

let queryCount = 0;
async function handleTeamsMessagingExtensionQuery(
    context: TurnContext,
    query: MessagingExtensionQuery
): Promise<MessagingExtensionResponse> {

    let revenueRange = cleanupParam(query.parameters[0]?.value);

    console.log(`💰 Revenue query #${++queryCount}: Products with revenue in range=${revenueRange}`);

    const products = await getProductsByRevenueRange(revenueRange);

    console.log(`Found ${products.length} products in the Northwind database`)
    const attachments = [];
    products.forEach((product) => {
        const preview = CardFactory.heroCard(product.ProductName,
            `Revenue/period ${product.Revenue.toLocaleString()}`, [product.ImageUrl]);

        const resultCard = cardHandler.getEditCard(product);
        const attachment = { ...resultCard, preview };
        attachments.push(attachment);
    });
    return {
        composeExtension: {
            type: "result",
            attachmentLayout: "list",
            attachments: attachments,
        },
    };
}

function cleanupParam(value: string): string {

    if (!value) {
        return "";
    } else {
        let result = value.trim();
        result = result.split(',')[0];          // Remove extra data
        result = result.replace("*", "");       // Remove wildcard characters from Copilot
        return result;
    }
}

export default { COMMAND_ID, handleTeamsMessagingExtensionQuery }

File: samples/msgext-northwind-inventory-action-csharp/MessageExtensions/RevenueSearchCommand.cs
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Schema.Teams;
using NorthwindInventory.NorthwindDB;
using NorthwindInventory.Handlers;

namespace NorthwindInventory.MessageExtensions
{
    public static class RevenueSearchCommand
    {
        public const string CommandId = "revenueSearch";
        private static int queryCount = 0;

        public static async Task<MessagingExtensionResponse> HandleTeamsMessagingExtensionQueryAsync(
            ITurnContext<IInvokeActivity> turnContext,
            MessagingExtensionQuery query,
            IConfiguration configuration,
            CancellationToken cancellationToken)
        {
            // Seek the parameter by name, don't assume it's in element 0 of the array
            string revenueRange = Utils.CleanupParam(
                query.Parameters?.FirstOrDefault(p => p.Name == "revenueRange")?.Value as string);

            Console.WriteLine($"💰 Revenue query #{++queryCount}: Products with revenue in range={revenueRange}");

            var productService = new ProductService(configuration);
            var products = await productService.GetProductsByRevenueRange(revenueRange);

            Console.WriteLine($"Found {products.Count} products in the Northwind database");
            var attachments = new List<MessagingExtensionAttachment>();

            foreach (var product in products)
            {
                var preview = new HeroCard
                {
                    Title = product.ProductName,
                    Subtitle = $"Revenue/period {product.Revenue.ToString("N0")}",
                    Images = new List<CardImage> { new(product.ImageUrl) }
                }.ToAttachment();

                var resultCard = CardHandler.GetEditCard(product);

                var attachment = new MessagingExtensionAttachment
                {
                    ContentType = resultCard.ContentType,
                    Content = resultCard.Content,
                    Preview = preview
                };

                attachments.Add(attachment);
            }

            return new MessagingExtensionResponse
            {
                ComposeExtension = new MessagingExtensionResult
                {
                    Type = "result",
                    AttachmentLayout = "list",
                    Attachments = attachments
                }
            };
        }
    }
}